cuqi/density/_density.py

```diff
@@ -0,0 +1,113 @@
+
+    Parameters
+    ----------
+    name: str
+        Name of the random variable associated with the density.
+
+    Notes
+    -----
+    Subclasses must implement the following:
+
+    - _logd(self, *args, **kwargs). Returns the log density given a set of parameters.
+    - _condition(self, *args, **kwargs). This method conditions (fixes) the density on a given set of parameters.
+    - dim. The dimension of the density.
+    - get_parameter_names(self). Returns a list of the names of the parameters of the density.
+
+    """
+    def __init__(self, name: Optional[str] = None):
+        if not isinstance(name, str) and name is not None:
+            raise ValueError("Name must be a string or None")
+        self.name = name
+        self._constant = 0 # Precomputed constant to add to the log probability.
+
+    def logd(self, *args, **kwargs):
+        """ Evaluates the log density given a set of parameters.
+        
+        The log density can be evaluated with a set of arguments matching the parameter names
+        of the density. These can be accessed with the :meth:`get_parameter_names` method.
+        """
+
+        # Get parameter names possible to evaluate the logd
+        par_names = self.get_parameter_names()
+
+        # Check if kwargs are given. If so parse them according to the parameter names and add them to args.
+        if len(kwargs) > 0:
+
+            if len(args) > 0:
+                raise ValueError("Density.logd: Cannot specify both positional and keyword arguments.")
+
+            # Check if parameter names match the keyword arguments (any order).
+            if set(par_names) != set(kwargs.keys()):
+                raise ValueError(f"Density.logd: Parameter names {par_names} do not match keyword arguments {kwargs.keys()}.")
+            
+            # Ensure that the keyword arguments are given in the correct order and use them as positional arguments.
+            args = [kwargs[name] for name in par_names]            
+
+        # Check if the number of arguments matches the number of parameters.
+        if len(args) != len(par_names):
+            raise ValueError(f"Density.logd: Number of arguments must match number of parameters. Got {len(args)} arguments but density has {len(par_names)} parameters.")
+
+        return self._logd(*args) + self._constant
+   
+    @abstractmethod
+    def _logd(self):
+        pass
+
+    @abstractmethod
+    def _condition(self):
+        pass
+
+    @property
+    @abstractmethod
+    def dim(self):
+        pass
+
+    @abstractmethod
+    def get_parameter_names(self):
+        """ Returns the names of the parameters that the density can be evaluated at or conditioned on. """
+        pass
+
+    def __call__(self, *args, **kwargs) -> Union[cuqi.distribution.Distribution, cuqi.likelihood.Likelihood, ConstantDensity]:
+        """ Condition on the given parameters. """
+        return self._condition(*args, **kwargs)
+
+class ConstantDensity(Density):
```

Comment:
Hmm. I am unsure about good naming here. I will leave this for @jakobsj to weigh in on.

Reply:
This is tricky. I agree that `ConstantDensity` is potentially confusing (it is to me). But what is the right/best name?  What would the mathematical term be? Other suggestions that come mind: ObservedDensity and SingularDensity (but the latter may have a specific unrelated meaning).

Reply:
As subclassed from `Density` it remains a density and as such I would say it would be good to end with `Density`.

Reply:
I think `EvaluatedDensity` is maybe the most clear. It is essentially just a single number (since the densities map to R), but it keeps the same interface as the remaning densities (name, logd, dim etc.)

It is a density function, that is evaluated. Hence it simply returns the value of the evaluated density and the parameter input dimension is zero since there are not parameters for an evaluated density.